### ISSUE TITLE
Databend Roadmap for 2024 (Discussion)

### ISSUE THREAD

**Issue by BohuTANG** [datafuselabs/databend#14167]
# Databend Roadmap for 2024 (Discussion)

Explore our ongoing journey and future plans for Databend. Join the discussion and contribute your ideas!

## Review of 2023

In 2023, Databend scaled significantly. The largest single table in Databend managed to handle hundreds of thousands of segments, several ten million blocks, tens of trillions of records, encompassing 7PB of raw data and over 300TB of index data.

## Main Tasks for 2024

Here's a snapshot of the key tasks we're focusing on:

| Task                                      | Status        | Comments                                                    |
|-------------------------------------------|---------------|-------------------------------------------------------------|
| Enhancements to Concurrency and Scheduler | Planned       | Aiming for faster, more efficient task handling and improved system responsiveness. |
| TPC-DS Performance                        | In Progress   | Continuously optimizing for better performance benchmarks.   |
| Support Python Worksheet                  | Not Specified | Adding Python support for versatile data analysis alongside SQL. |
| Unify Storage, Warehouse, and Compute     | Not Specified | Creating a cohesive data platform for AI and cloud computing, provisioning CPU & GPU resources. |

## Previous Roadmaps for Reference:
- **2023 Roadmap:** [Issue #9448](https://github.com/datafuselabs/databend/issues/9448)
- **2022 Roadmap:** [Issue #3706](https://github.com/datafuselabs/databend/issues/3706)
- **2021 Roadmap:** [Issue #746](https://github.com/datafuselabs/databend/issues/746)

**Comment by GaoYusong** [comment]
Congratulations on what Databend has achieved in such a short time. Looking forward to 2024!

**Comment by Xuanwo** [comment]
So what's `Support Python Worksheet`, is it means running python in databend?

**Comment by ZhiHanZ** [comment]
Any plan for SQL transaction and stored procedures?

**Comment by ZhiHanZ** [comment]
besides, I think we could also support query queueing, warehouse automatica scaling based on pending queue and separate another coordinator component for dispatching physical plan to warehouse compute node.

**Comment by BohuTANG** [comment]
> besides, I think we could also support query queueing, warehouse automatica scaling based on pending queue and separate another coordinator component for dispatching physical plan to warehouse compute node.

This is a part of `Enhancements to Concurrency and Scheduler`.

**Comment by BohuTANG** [comment]
> What exactly is the `Support Python Worksheet`? Does it enable running Python in Databend?

The goal is to make the Hugging Face Model + Python + GPU( or CPU) + Data in Databend is possible.

**Comment by djouallah** [comment]
all I want is to be able to read a delta table from a local path :)

**Comment by JasonLi-cn** [comment]
How to understand **Inference** ？Which abilities does it refer to?

**Comment by BohuTANG** [comment]
> How to understand **Inference** ？Which abilities does it refer to?

Move the models(huggingface models) to the database, the database can load and run them.

**Comment by keltia** [comment]
Thanks for making this available to everyone.  I'm currently interested to anything that can deal with Geospatial stuff so adding Geometry support is very nice.

**Comment by BohuTANG** [comment]
> Thanks for making this available to everyone. I'm currently interested to anything that can deal with Geospatial stuff so adding Geometry support is very nice.

Working on it: https://github.com/datafuselabs/databend/pull/14470

**Comment by inviscid** [comment]
> Thanks for making this available to everyone. I'm currently interested to anything that can deal with Geospatial stuff so adding Geometry support is very nice.

@keltia Databend uses H3 for geospatial operations.  Is that what you are referring to? https://docs.databend.com/sql/sql-functions/geo-functions/

**Comment by inviscid** [comment]
Would it be possible to restructure the code a little to help with creating a fully OSS compliant version of Databend?

I know the license info specifically calls out the `ee` directories but we have found some other files that are also covered by the Elastic license that seem to be part of the core functionality. 
 - `/src/meta/binaries/meta/ee-main.rs`
 - `/src/binaries/query/ee-main.rs`

Many companies will not adopt a technology that doesn't use an approved OSS license, which Apache 2.0 is, but Elastic is not (open-source vs. source-available).

In an ideal world there would be a `databend-oss` repo and a `databend-ee` repo with the latter adding all of the Enterprise features and licensing on top of the OSS version.

I know the goal is to get enterprise customers to buy a license but I think you may be losing out on more community support and growth because devs restricted by corporate governance policies will never be able to even give Databend a test drive.  Plus, like many grassroots database projects within a company, things start small and cheap then turn into mission critical systems requiring licensing.  If we skip the small and cheap phase, the licensing never comes in those cases.

**Comment by johnpyp** [comment]
Agree, the mixing and matching of licenses both in docs conceptually and in the literal code makes it really hard to be able to use this database.

**Comment by cdmikechen** [comment]
@BohuTANG 
Hope to add support for spatial indexing of geometry in 2025 (like https://github.com/opengeospatial/geoparquet/issues/13 ), as well as optimize too many snapshots problem for high-frequency time series data storage.

**Comment by Xuanwo** [comment]
I wish databend in 2025 can support all of iceberg, paimon and hudi.

**Comment by BohuTANG** [comment]
> @BohuTANG Hope to add support for spatial indexing of geometry in 2025 (like [opengeospatial/geoparquet#13](https://github.com/opengeospatial/geoparquet/issues/13) ), as well as optimize too many snapshots problem for high-frequency time series data storage.

Added to the roadmap. Thank you for the valuable suggestions!

**Comment by BohuTANG** [comment]
> I wish databend in 2025 can support all of iceberg, paimon and hudi.

Thanks! This goal has been added to the roadmap.

**Comment by wubx** [comment]
Dynamic tables are useful in many scenarios.

**Comment by zhanglistar** [comment]
What about streaming computation support, like time window, watermark and stateful functions?

**Comment by BohuTANG** [comment]
> What about streaming computation support, like time window, watermark and stateful functions?

No plan yet. Could you give more about this  requirements background?

**Comment by zhanglistar** [comment]
> > What about streaming computation support, like time window, watermark and stateful functions?
> 
> No plan yet. Could you give more about this requirements background?

Not so much background, just a new direction of combine stream computinng and batch computing together.
